hal/explicit_pinmap.h

```diff
@@ -0,0 +1,180 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef EXPLICIT_PINMAP_H
+#define EXPLICIT_PINMAP_H
+#include "PeripheralPinMaps.h"
+
+MSTD_CONSTEXPR_FN_14 PinMap get_pwm_pinmap(const PinName pin)
+{
+    for (const PinMap &pinmap : PINMAP_PWM) {
+        if (pinmap.pin == pin) {
+            return {pin, pinmap.peripheral, pinmap.function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+MSTD_CONSTEXPR_FN_14 PinMap get_analogin_pinmap(const PinName pin)
+{
+    for (const PinMap &pinmap : PINMAP_ANALOGIN) {
+        if (pinmap.pin == pin) {
+            return {pin, pinmap.peripheral, pinmap.function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+MSTD_CONSTEXPR_FN_14 PinMap get_analogout_pinmap(const PinName pin)
+{
+    for (const PinMap &pinmap : PINMAP_ANALOGIN) {
+        if (pinmap.pin == pin) {
+            return {pin, pinmap.peripheral, pinmap.function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+MSTD_CONSTEXPR_FN_14 i2c_pinmap_t get_i2c_pinmap(const PinName sda, const PinName scl)
+{
+    int sda_idx = NC;
+    int scl_idx = NC;
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_I2C_SDA)/sizeof(PINMAP_I2C_SDA[0])); i++) {
+        if (PINMAP_I2C_SDA[i].pin == sda) {
+            sda_idx = i;
+            break;
+        }
+    }
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_I2C_SCL)/sizeof(PINMAP_I2C_SCL[0])); i++) {
+        if (PINMAP_I2C_SCL[i].pin == scl) {
+            scl_idx = i;
+            break;
+        }
+    }
+
+    if (sda_idx == NC || scl_idx == NC || PINMAP_I2C_SDA[sda_idx].peripheral != PINMAP_I2C_SCL[scl_idx].peripheral) {
+        return {NC, NC, NC, NC, NC};
+    }
+
+    return {PINMAP_I2C_SDA[sda_idx].peripheral, PINMAP_I2C_SDA[sda_idx].pin, PINMAP_I2C_SDA[sda_idx].function, PINMAP_I2C_SCL[scl_idx].pin, PINMAP_I2C_SCL[scl_idx].function};
+}
+
+MSTD_CONSTEXPR_FN_14 serial_pinmap_t get_uart_pinmap(const PinName tx, const PinName rx)
+{
+    int tx_idx = NC;
+    int rx_idx = NC;
+
+    for (uint32_t i = 0; i < (sizeof(PINMAP_UART_TX)/sizeof(PINMAP_UART_TX[0])); i++) {
```

Comment:
If you wanted to extend these to the ranged for, it would be

    const PinMap *tx_map = nullptr;
    for (const PinMap &pinmap : PINMAP_UART_TX) {
        if (pinmap.pin == tx) {
            tx_map = &pinmap;
            break;
        }
    }

But at that point, the `std::find_if` version is actually more concise:

    const PinMap *tx_map = std::find_if(std::begin(PINMAP_UART_TX), std::end(PINMAP_UART_TX),
                                        [tx](const PinMap &map) { return map.pin == tx; });
    const PinMap *rx_map = std::find_if(std::begin(PINMAP_UART_RX), std::end(PINMAP_UART_RX),
                                        [rx](const PinMap &map) { return map.pin == rx; });

    if (!tx_map || !rx_map || tx_map->peripheral != rx_map->peripheral) {
        return {NC, NC, NC, NC, NC};
    }

    return {pinmap_tx->peripheral, pinmap_tx->pin, pinmap_tx->function, pinmap_rx->pin, pinmap_tx->function};

Still a bit clunky, so you could factor it out to a helper:

    template<size_t N>
    constexpr const PinMap *pinmap_lookup_pin(const PinMap(&maptable)[N], PinName pin)
    {
        // or use the range-loop here if you prefer. Important point was passing
        // the array by reference to capture its size for the iteration.
        return std::find_if(std::begin(maptable), std::end(maptable),
                            [pin](const PinMap &map) { return map.pin == pin; });
    }

    const PinMap *tx_map = pinmap_lookup_pin(PINMAP_UART_TX, tx);
    const PinMap *rx_map = pinmap_lookup_pin(PINMAP_UART_RX, rx);

Reply:
Actually, I didn't realise that `std::find_if` isn't `constexpr` in C++14 - you'll have to use the ranged-for loop. The helper function would still make it neater though.